File: pronunciations/js/example_pronunciation_dictionaries.js
#!/usr/bin/env node
/** Manage a named pronunciation dictionary through Inworld's public REST API. */

try {
    require('dotenv').config();
} catch (_) {}

const DEFAULT_API_BASE_URL = 'https://api.inworld.ai';

const INITIAL_PRONUNCIATIONS = [
    {
        displayHeadword: 'Cat',
        languageCode: 'en-US',
        phoneSymbols: ['k', 'æ', 't'],
    },
    {
        displayHeadword: 'Ship',
        languageCode: 'en-US',
        phoneSymbols: ['ʃ', 'ɪ', 'p'],
    },
];

const REPLACEMENT_PRONUNCIATIONS = [
    {
        displayHeadword: 'Red',
        languageCode: 'en-US',
        phoneSymbols: ['ɹ', 'ɛ', 'd'],
    },
];

class PronunciationDictionariesClient {
    constructor(apiKey, workspaceId, apiBaseUrl) {
        this.apiKey = apiKey;
        this.workspaceId = workspaceId;
        this.apiBaseUrl = apiBaseUrl.replace(/\/$/, '');
        this.collectionPath =
            `/pronunciations/v1/workspaces/${encodeURIComponent(workspaceId)}` +
            '/pronunciationDictionaries';
    }

    createDictionary(displayName, pronunciations) {
        return this.request(this.collectionPath, {
            method: 'POST',
            body: { displayName, pronunciations },
        });
    }

    listDictionaries(pageSize = 5) {
        return this.request(this.collectionPath, {
            query: { pageSize: String(pageSize) },
        });
    }

    getDictionary(name) {
        return this.request(this.resourcePath(name));
    }

    updateDictionary(dictionary, displayName, pronunciations) {
        return this.request(this.resourcePath(dictionary.name), {
            method: 'PATCH',
            query: { updateMask: 'displayName,pronunciations' },
            body: {
                name: dictionary.name,
                displayName,
                pronunciations,
                etag: dictionary.etag,
            },
        });
    }

    deleteDictionary(dictionary) {
        return this.request(this.resourcePath(dictionary.name), {
            method: 'DELETE',
            query: { etag: dictionary.etag },
        });
    }

    resourcePath(name) {
        const expectedPrefix =
            `workspaces/${this.workspaceId}/pronunciationDictionaries/`;
        if (!name.startsWith(expectedPrefix)) {
            throw new Error(`Unexpected pronunciation dictionary name: ${name}`);
        }
        const encodedName = name.split('/').map(encodeURIComponent).join('/');
        return `/pronunciations/v1/${encodedName}`;
    }

    async request(path, { method = 'GET', query = {}, body } = {}) {
        const url = new URL(`${this.apiBaseUrl}${path}`);
        for (const [name, value] of Object.entries(query)) {
            url.searchParams.set(name, value);
        }

        const response = await fetch(url, {
            method,
            headers: {
                Authorization: `Basic ${this.apiKey}`,
                ...(body ? { 'Content-Type': 'application/json' } : {}),
            },
            ...(body ? { body: JSON.stringify(body) } : {}),
        });
        const responseText = await response.text();
        if (!response.ok) {
            throw new Error(
                `${method} ${url} failed with HTTP ${response.status}: ${responseText}`,
            );
        }
        return responseText ? JSON.parse(responseText) : {};
    }
}

function printResult(label, value) {
    console.log(`\n${label}`);
    console.log(JSON.stringify(value, null, 2));
}

function requireEnvironmentVariable(name) {
    const value = process.env[name];
    if (!value) {
        throw new Error(
            `${name} is not set. Copy .env.example to .env and provide a value.`,
        );
    }
    return value;
}

async function runLifecycle(client) {
    let dictionary = null;
    const timestamp = new Date().toISOString().replace(/[:.]/g, '-');

    try {
        dictionary = await client.createDictionary(
            `API example ${timestamp}`,
            INITIAL_PRONUNCIATIONS,
        );
        printResult('1. Created dictionary', dictionary);

        const page = await client.listDictionaries(5);
        printResult('2. Listed dictionaries (first page)', page);

        dictionary = await client.getDictionary(dictionary.name);
        printResult('3. Retrieved dictionary', dictionary);

        dictionary = await client.updateDictionary(
            dictionary,
            `API example updated ${timestamp}`,
            REPLACEMENT_PRONUNCIATIONS,
        );
        printResult('4. Replaced dictionary contents', dictionary);

        const deleteResponse = await client.deleteDictionary(dictionary);
        printResult('5. Deleted dictionary', deleteResponse);
        dictionary = null;
    } finally {
        if (dictionary) {
            try {
                await client.deleteDictionary(dictionary);
                console.log('\nCleaned up the dictionary after an incomplete run.');
            } catch (cleanupError) {
                console.error(
                    `\nCleanup failed; delete ${dictionary.name} manually: ` +
                    cleanupError.message,
                );
            }
        }
    }
}

async function main() {
    const client = new PronunciationDictionariesClient(
        requireEnvironmentVariable('INWORLD_API_KEY'),
        requireEnvironmentVariable('INWORLD_WORKSPACE_ID'),
        process.env.INWORLD_API_BASE_URL || DEFAULT_API_BASE_URL,
    );
    await runLifecycle(client);
}

main().catch((error) => {
    console.error(`Pronunciation dictionary example failed: ${error.message}`);
    process.exitCode = 1;
});

File: pronunciations/python/example_pronunciation_dictionaries.py
#!/usr/bin/env python3
"""Manage a named pronunciation dictionary through Inworld's public REST API."""

import json
import os
import sys
from datetime import datetime, timezone
from typing import Any
from urllib.parse import quote

try:
    from dotenv import load_dotenv

    load_dotenv()
except ImportError:
    pass

import requests

DEFAULT_API_BASE_URL = "https://api.inworld.ai"
REQUEST_TIMEOUT_SECONDS = 30

INITIAL_PRONUNCIATIONS = [
    {
        "displayHeadword": "Cat",
        "languageCode": "en-US",
        "phoneSymbols": ["k", "æ", "t"],
    },
    {
        "displayHeadword": "Ship",
        "languageCode": "en-US",
        "phoneSymbols": ["ʃ", "ɪ", "p"],
    },
]

REPLACEMENT_PRONUNCIATIONS = [
    {
        "displayHeadword": "Red",
        "languageCode": "en-US",
        "phoneSymbols": ["ɹ", "ɛ", "d"],
    }
]


class PronunciationDictionariesClient:
    """Minimal client for the five public pronunciation dictionary methods."""

    def __init__(self, api_key: str, workspace_id: str, api_base_url: str):
        self.api_base_url = api_base_url.rstrip("/")
        self.workspace_id = workspace_id
        encoded_workspace_id = quote(workspace_id, safe="")
        self.collection_url = (
            f"{self.api_base_url}/pronunciations/v1/workspaces/"
            f"{encoded_workspace_id}/pronunciationDictionaries"
        )
        self.session = requests.Session()
        self.session.headers.update(
            {
                "Authorization": f"Basic {api_key}",
                "Content-Type": "application/json",
            }
        )

    def create_dictionary(
        self, display_name: str, pronunciations: list[dict[str, Any]]
    ) -> dict[str, Any]:
        return self._request(
            "POST",
            self.collection_url,
            json={
                "displayName": display_name,
                "pronunciations": pronunciations,
            },
        )

    def list_dictionaries(self, page_size: int = 5) -> dict[str, Any]:
        return self._request(
            "GET",
            self.collection_url,
            params={"pageSize": page_size},
        )

    def get_dictionary(self, name: str) -> dict[str, Any]:
        return self._request("GET", self._resource_url(name))

    def update_dictionary(
        self,
        dictionary: dict[str, Any],
        display_name: str,
        pronunciations: list[dict[str, Any]],
    ) -> dict[str, Any]:
        return self._request(
            "PATCH",
            self._resource_url(dictionary["name"]),
            params={"updateMask": "displayName,pronunciations"},
            json={
                "name": dictionary["name"],
                "displayName": display_name,
                "pronunciations": pronunciations,
                "etag": dictionary["etag"],
            },
        )

    def delete_dictionary(self, dictionary: dict[str, Any]) -> dict[str, Any]:
        return self._request(
            "DELETE",
            self._resource_url(dictionary["name"]),
            params={"etag": dictionary["etag"]},
        )

    def _resource_url(self, name: str) -> str:
        expected_prefix = f"workspaces/{self.workspace_id}/pronunciationDictionaries/"
        if not name.startswith(expected_prefix):
            raise ValueError(f"Unexpected pronunciation dictionary name: {name}")
        encoded_name = "/".join(quote(segment, safe="") for segment in name.split("/"))
        return f"{self.api_base_url}/pronunciations/v1/{encoded_name}"

    def _request(self, method: str, url: str, **kwargs: Any) -> dict[str, Any]:
        response = self.session.request(
            method,
            url,
            timeout=REQUEST_TIMEOUT_SECONDS,
            **kwargs,
        )
        if not response.ok:
            raise RuntimeError(
                f"{method} {response.url} failed with HTTP {response.status_code}: "
                f"{response.text}"
            )
        return response.json()


def print_result(label: str, value: dict[str, Any]) -> None:
    """Print an API response without escaping IPA symbols."""
    print(f"\n{label}")
    print(json.dumps(value, indent=2, ensure_ascii=False))


def run_lifecycle(client: PronunciationDictionariesClient) -> None:
    """Run create, list, get, update, and delete with cleanup on failure."""
    dictionary: dict[str, Any] | None = None
    timestamp = datetime.now(timezone.utc).strftime("%Y%m%d-%H%M%S")

    try:
        dictionary = client.create_dictionary(
            display_name=f"API example {timestamp}",
            pronunciations=INITIAL_PRONUNCIATIONS,
        )
        print_result("1. Created dictionary", dictionary)

        page = client.list_dictionaries(page_size=5)
        print_result("2. Listed dictionaries (first page)", page)

        dictionary = client.get_dictionary(dictionary["name"])
        print_result("3. Retrieved dictionary", dictionary)

        dictionary = client.update_dictionary(
            dictionary=dictionary,
            display_name=f"API example updated {timestamp}",
            pronunciations=REPLACEMENT_PRONUNCIATIONS,
        )
        print_result("4. Replaced dictionary contents", dictionary)

        delete_response = client.delete_dictionary(dictionary)
        print_result("5. Deleted dictionary", delete_response)
        dictionary = None
    finally:
        if dictionary is not None:
            try:
                client.delete_dictionary(dictionary)
                print("\nCleaned up the dictionary after an incomplete run.")
            except (
                requests.RequestException,
                RuntimeError,
                ValueError,
                KeyError,
            ) as cleanup_error:
                print(
                    f"\nCleanup failed; delete {dictionary['name']} manually: "
                    f"{cleanup_error}",
                    file=sys.stderr,
                )


def require_environment_variable(name: str) -> str:
    """Return a required environment variable or raise a focused error."""
    value = os.getenv(name)
    if not value:
        raise ValueError(
            f"{name} is not set. Copy .env.example to .env and provide a value."
        )
    return value


def main() -> int:
    """Load configuration and run the example."""
    try:
        client = PronunciationDictionariesClient(
            api_key=require_environment_variable("INWORLD_API_KEY"),
            workspace_id=require_environment_variable("INWORLD_WORKSPACE_ID"),
            api_base_url=os.getenv("INWORLD_API_BASE_URL", DEFAULT_API_BASE_URL),
        )
        run_lifecycle(client)
    except (requests.RequestException, RuntimeError, ValueError, KeyError) as error:
        print(f"Pronunciation dictionary example failed: {error}", file=sys.stderr)
        return 1

    return 0


if __name__ == "__main__":
    sys.exit(main())
